### ISSUE TITLE
OR operator or a way to merge different searches into one

### ISSUE THREAD

**Issue by elreymu** [Bionus/imgbrd-grabber#493]
This is done in danbooru by using the ~ prefix but it's not working with gelbooru

For example i want to search for two or more different characters at the same time and have the individual results show up in the same tab
search: ~tag1 ~tag2
would show up posts that have tag1 OR tag2.
alternatively, having a way to merge two different searches

**Comment by Bionus** [comment]
Hello, sorry for the late reply.
It can be done (however not easily at all in the current state of things),
but I actually never understood the need for this operator.

What's the advantage of doing "~tag1 ~tag2" over doing "tag1" then "tag2"
in a later search?
Convenience of having both results as the same time? Doesn't the fact that
there are multiple pages of results cancels this advantage since you have
to browse through multiple pages/searches anyway?
I'm asking in order to be sure I'm not missing/misunderstanding anything
before starting.

**Comment by elreymu** [comment]
@Bionus 
Well actually the advantage is pretty huge if used with creativity.
For example, it's useful when looking for 2 different tags that are similar (and both of them give similar results) 
Such as "Solo" and "1girl" or "1male" when i am looking for pictures of a certain character.
with having the "~tag1 ~tag2" i don't have to do several searches, specially if i am looking for pictures of other characters at the same time.
Another useful thing is that i could search for my favorite artists  while at the same time searching a series in particular.
It would be something like "~autor1 ~autor2 ~autor3 ~touhou ~kantai_collection ~capcom"
It would be extremely helpful.
It also would be helpful when looking for new high score pictures with specific traits or tags " ~tag1 ~tag2 ~tag3 ~tag4 ~tag5 ~tag6 score:>9" and have everything show up in the same tab

**Comment by Meem0** [comment]
@Bionus @elreymu 

Yes this is actually a really deep topic, and one that I've spent a lot of time thinking about.

OR operator allows you search for high-level concepts that don't have a single tag, or are spread across multiple tags.

Some examples:
- If you want some nature in your picture, you can add (~landscape ~scenic ~tree ~flower ~plant ...) (not sure about what the standard tags are there).
- If you want a scene in a room, you can add (~bed ~pillow ~chair ~window ...) to your search.
- If you want a fantasy-type character (like out of an RPG), you can add things like (~fantasy ~pointed_ears ~staff ~tiara ~circlet ~wand ~side_slit ...).

Then you can AND together those groups to get things like a fantasy-type character in a nature-filled scene, and so on.

IbSearch has this functionality already, but unfortunately the website is very slow.  [Here is a link to a search on IbSear.ch](https://ibsear.ch/images/?q=%28armor+%7C+circlet+%7C+fantasy+%7C+pointed_ears+%7C+side_slit+%7C+staff+%7C+tiara+%7C+wand+%7C+weapon%29+%28cloud+%7C+grass+%7C+sky+%7C+tree%29+-forest) (SFW only), which gives a quick example of using combinations of OR groups to get "fantasy characters in a natural / scenic setting, but not a forest" which would be a hard concept to pin down without the OR functionality.


Now, if you don't mind me advancing this topic a few big steps forward: I'm also very interested in investigating "weighted tag searching."  By this I mean associating a "point" value to a number of tags or OR-groups of tags, then returning images that have the highest scores.  This allows you to say "it would be nice if the image had all of these properties, but I don't want to narrow it down, so get as many as you can."

Consider a search like "japanese_clothes{40} water{15} mountain{25} sunset{20} (~tree ~grass ~flower ~nature){20} absurdres{35} ..." (example syntax).  You're saying it would be nice to have all of those tags, but it's okay if you miss a few; and also that some are more important than others.

For another example, look at my previous IbSearch from before.  Since I OR'd together cloud/sky/grass/nature, etc., it's actually possible to get a desolate landscape with a big open sky, which might not be what I want.  With weighted tag searching, I could have put a high weight on the nature tags, then a lower weight on cloud and sky.

You really don't have to look much further than the absurdres tag to see why this would be so valuable: I rarely want to restrict my search to only posts with absurdres, but 100% of the time I would prefer to have images tagged with it.

I wonder how feasible this is to implement.  It's super interesting to me, and I'm also learning about a very similar topic in my AI course right now (local search / constraint satisfaction), so I would be happy to do what I can to contribute, if it's deemed a worthwhile feature.  If not, I'll probably make my own branch.
